### PR TITLE
Refactor wharfyml, moved out Docker PodSpec creation

### PR DESCRIPTION
**Pull request by jilleJr**
- \[x] I've added a new note in the `CHANGELOG.md` file, according to docs:
  https://iver-wharf.github.io/#/development/changelogs/writing-changelogs

## Summary

- Split out code from wharfyml into new packages:

  - `pkg/wharfyml.Errors` -> `internal/errutil.Slice`
  - `pkg/wharfyml.PathError` -> `internal/errutil.Scoped`
  - `pkg/wharfyml.PosError` -> `internal/errutil.Pos`
  - `pkg/wharfyml.Pos` -> `pkg/wharfyml/visit.Pos`
  - `pkg/wharfyml.Step*` -> `pkg/steps/steps.*`
  - `pkg/wharfyml.nodeMapParser` -> `pkg/wharfyml/visit.MapVisitor`
  - `pkg/wharfyml/visitor.go` -> `pkg/wharfyml/visit/visit.go`

- Added new packages:

  - `internal/errtesting`
  - `internal/errutil`
  - `internal/yamltesting`
  - `pkg/steps`
  - `pkg/wharfyml/visit`

- Moved Docker `v1.PodSpec` templating out from `pkg/worker/k8spodtemplating.go` to `pkg/steps/docker.go`

- Changed package in `pkg/wharfyml/parse_test.go` from `wharfyml` to `wharfyml_test`, as it needs to depend on `pkg/steps`, and therefore cannot have circular dependency between `pkg/steps` and `pkg/wharfyml`.

- Changed `docker.secretArgs` and `docker.secretName` from runtime validation warnings to parse-time errors

## Motivation

This whole thing comes from the idea that:

- Step types should both prepare as much as possible as early as possible.
- Want to be able to use varsub when creating pods.
- The wharfyml package should not have the steps hardcoded.

I've not yet changed the remaining step types to only be defined inside `pkg/steps`. I though I'd be able to leave that for a future PR, to try make this huge PR not become huge'er.

This relates to #65 and #124, but does not close them yet. I need to move the remaining step types from `pkg/worker/k8spodtemplating.go` to `pkg/steps`

### CODE REVIEW

File: internal/yamltesting/yamltesting.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+	"fmt"
+	"testing"
+
+	"github.com/iver-wharf/wharf-cmd/pkg/wharfyml/visit"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"gopkg.in/yaml.v3"
+)
+
+func NewKeyedNode(t *testing.T, content string) (visit.StringNode, *yaml.Node) {
```
Comment (jilleJr) [inline internal/yamltesting/yamltesting.go:13]: I'm missing godoc on a lot of functions. Will fix.

**Review by Alexamakans — CHANGES_REQUESTED** [review]
Always nice with a good refactor 🎉

File: cmd/wharf/utils.go [hunk 2]
```diff
@@ -110,21 +115,27 @@ func parseInputArgs(inputs flagtypes.KeyValueArray) map[string]any {
...
+		scopePrefix := errutil.AsScope(err)
+		if scopePrefix != "" {
+			scopePrefix += ": "
+		}
+		var posErr errutil.Pos
 		if errors.As(err, &posErr) {
-			log.Warn().Messagef("%4d:%-4d%s",
-				posErr.Source.Line, posErr.Source.Column, err.Error())
+			log.Warn().Messagef("%4d:%-4d%s%s",
+				posErr.Line, posErr.Column, scopePrefix, err.Error())
```
[the remark this answers is not in the document]
Comment (jilleJr) [reply cmd/wharf/utils.go:129]: Ah yea might add another space just for safety :)

File: pkg/steps/docker.go [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+			MountPath: "/kaniko/.docker",
+		})
+	}
+
+	args := []string{
+		// Not using path/filepath package because we know don't want to
+		// suddenly use Windows directory separator when running from Windows.
+		"--dockerfile", path.Join(repoDir, s.File),
+		"--context", path.Join(repoDir, s.Context),
+		"--skip-tls-verify", // This is bad, but remains due to backward compatibility
```
[the remark this answers is not in the document]
Comment (jilleJr) [reply pkg/steps/docker.go]: Good idea!

File: pkg/steps/docker.go [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+		Image: fmt.Sprintf("%s:%s", s.config.Image, s.config.ImageTag),
+		// default entrypoint for image is "/kaniko/executor"
+		WorkingDir: repoDir,
+		VolumeMounts: []v1.VolumeMount{
+			commonRepoVolumeMount,
+		},
+	}
+
+	if s.AppendCert {
+		cont.VolumeMounts = append(cont.VolumeMounts,
```
[the remark this answers is not in the document]
Comment (jilleJr) [reply pkg/steps/docker.go:173]: Sure :)

File: internal/yamltesting/yamltesting.go (hunk 1 above)
Comment (jilleJr) [reply internal/yamltesting/yamltesting.go:15]: Added :disappointed_relieved:

File: pkg/wharfyml/step.go [hunk 5]
```diff
@@ -24,16 +24,16 @@ type Step struct {
...
 	errSlice.Add(errs...)
 	if len(nodes) == 0 {
-		errSlice.Add(visit.WrapPosErrorNode(ErrStepEmpty, node))
+		errSlice.Add(errutil.NewPosFromNode(ErrStepEmpty, node))
 		return
 	}
 	if len(nodes) > 1 {
-		errSlice.Add(visit.WrapPosErrorNode(ErrStepMultipleStepTypes, node))
+		errSlice.Add(errutil.NewPosFromNode(ErrStepMultipleStepTypes, node))
 		// Continue, its not a fatal issue
```
Comment (Alexamakans) [inline pkg/wharfyml/step.go:37]:
```suggestion
		// Continue, it's not a fatal issue
```

File: pkg/wharfyml/visit/mapvisitor.go [hunk 6]
```diff
@@ -24,90 +26,93 @@ func NewMapVisitor(parent *yaml.Node, nodes map[string]*yaml.Node, source varsub
...
 }
 
+// ParentPos returns the position of the map node's parent.
 func (p MapVisitor) ParentPos() Pos {
-	return NewPosNode(p.parent)
+	return NewPosFromNode(p.parent)
 }
 
+// ReadNodesPos returns a map of the positions for all of the nodes that has so
+// far been visited.
```
Comment (Alexamakans) [inline pkg/wharfyml/visit/mapvisitor.go:44]:
```suggestion
// ReadNodesPos returns a map of the positions of all the nodes that have been
// visited so far.
```

File: pkg/wharfyml/visit/mapvisitor.go [hunk 7]
```diff
@@ -24,90 +26,93 @@ func NewMapVisitor(parent *yaml.Node, nodes map[string]*yaml.Node, source varsub
...
-		return errutil.Scope(err, key)
-	}
-	*target = num
-	return nil
+	return visitNode(p, key, target, Float64)
 }
 
+// RequireNumberFromVarSub looks up a value from the predefined varsub.Source
+// and writes the values to the pointer. An error is returned if the looked up
+// variable is not found or on type errors.
```
Comment (Alexamakans) [inline pkg/wharfyml/visit/mapvisitor.go:65]:
```suggestion
// RequireNumberFromVarSub looks up a value from the predefined varsub.Source
// and writes the value to the pointer. An error is returned if the looked up
// variable is not found or on type errors.
```

File: pkg/wharfyml/visit/mapvisitor.go [hunk 8]
```diff
@@ -24,90 +26,93 @@ func NewMapVisitor(parent *yaml.Node, nodes map[string]*yaml.Node, source varsub
...
 func (p MapVisitor) RequireNumberFromVarSub(varLookup string, target *float64) error {
 	return requireFromVarSub(p, varLookup, target, p.VisitNumber)
 }
 
+// LookupNumberFromVarSub looks up a value from the predefined varsub.Source
+// and writes the values to the pointer. An error is returned on type error.
+// If the variable is not present, then nil is returned and the pointer is
+// untouched.
```
Comment (Alexamakans) [inline pkg/wharfyml/visit/mapvisitor.go:73]:
```suggestion
// LookupNumberFromVarSub looks up a value from the predefined varsub.Source
// and writes the value to the pointer. An error is returned on type error.
// If the variable is not present, then nil is returned and the pointer is
// untouched.
```

File: pkg/wharfyml/visit/mapvisitor.go [hunk 9]
```diff
@@ -24,90 +26,93 @@ func NewMapVisitor(parent *yaml.Node, nodes map[string]*yaml.Node, source varsub
...
-		return errutil.Scope(err, key)
-	}
-	*target = num
-	return nil
+	return visitNode(p, key, target, Int)
 }
 
+// RequireIntFromVarSub looks up a value from the predefined varsub.Source
+// and writes the values to the pointer. An error is returned if the looked up
+// variable is not found or on type errors.
```
Comment (Alexamakans) [inline pkg/wharfyml/visit/mapvisitor.go:87]:
```suggestion
// RequireIntFromVarSub looks up a value from the predefined varsub.Source
// and writes the value to the pointer. An error is returned if the looked up
// variable is not found or on type errors.
```

File: pkg/wharfyml/visit/mapvisitor.go [hunk 10]
```diff
@@ -24,90 +26,93 @@ func NewMapVisitor(parent *yaml.Node, nodes map[string]*yaml.Node, source varsub
...
 func (p MapVisitor) RequireIntFromVarSub(varLookup string, target *int) error {
 	return requireFromVarSub(p, varLookup, target, p.VisitInt)
 }
 
+// LookupIntFromVarSub looks up a value from the predefined varsub.Source
+// and writes the values to the pointer. An error is returned on type error.
+// If the variable is not present, then nil is returned and the pointer is
+// untouched.
```
Comment (Alexamakans) [inline pkg/wharfyml/visit/mapvisitor.go:95]:
```suggestion
// LookupIntFromVarSub looks up a value from the predefined varsub.Source
// and writes the value to the pointer. An error is returned on type error.
// If the variable is not present, then nil is returned and the pointer is
// untouched.
```

File: pkg/wharfyml/visit/mapvisitor.go [hunk 11]
```diff
@@ -126,12 +131,15 @@ func (p MapVisitor) VisitStringSlice(key string, target *[]string) errutil.Slice
 	return errSlice
 }
 
+// VisitStringStringMap reads a node by string key and writes the string items
+// to the pointer. A slice of error contains any type errors. If the node is not
+// present, then nil is returned and the pointer is untouched.
```
Comment (Alexamakans) [inline pkg/wharfyml/visit/mapvisitor.go:136]: I feel this is clearer, I first thought of items as being just the value.
```suggestion
// VisitStringStringMap reads a node by string key and writes the string
// key-value pairs to the pointer. A slice of error contains any type errors. If
// the node is not present, then nil is returned and the pointer is untouched.
```

File: pkg/wharfyml/visit/mapvisitor.go [hunk 12]
```diff
@@ -149,6 +157,9 @@ func (p MapVisitor) VisitStringStringMap(key string, target *map[string]string)
 	return errSlice
 }
 
+// VisitStringWithVarSub will try find a variable from the predefined
+// varsub.Source and use that as a default if the node at the given key is not
+// present. An error is returned on type errors.
```
Comment (Alexamakans) [inline pkg/wharfyml/visit/mapvisitor.go:162]:
```suggestion
// VisitStringWithVarSub will try to find a variable from the predefined
// varsub.Source and use that as a default if the node at the given key is not
// present. An error is returned on type errors.
```

File: pkg/wharfyml/visit/mapvisitor.go [hunk 13]
```diff
@@ -157,36 +168,45 @@ func (p MapVisitor) VisitStringWithVarSub(nodeKey, varLookup string, target *str
 	return p.VisitString(nodeKey, target)
 }
 
+// RequireStringFromVarSub looks up a value from the predefined varsub.Source
+// and writes the values to the pointer. An error is returned if the looked up
+// variable is not found or on type errors.
```
Comment (Alexamakans) [inline pkg/wharfyml/visit/mapvisitor.go:173]:
```suggestion
// RequireStringFromVarSub looks up a value from the predefined varsub.Source
// and writes the value to the pointer. An error is returned if the looked up
// variable is not found or on type errors.
```

File: pkg/wharfyml/visit/mapvisitor.go [hunk 14]
```diff
@@ -157,36 +168,45 @@ func (p MapVisitor) VisitStringWithVarSub(nodeKey, varLookup string, target *str
...
 func (p MapVisitor) RequireStringFromVarSub(varLookup string, target *string) error {
 	return requireFromVarSub(p, varLookup, target, p.VisitString)
 }
 
+// LookupStringFromVarSub looks up a value from the predefined varsub.Source
+// and writes the values to the pointer. An error is returned on type error.
+// If the variable is not present, then nil is returned and the pointer is
+// untouched.
```
Comment (Alexamakans) [inline pkg/wharfyml/visit/mapvisitor.go:181]:
```suggestion
// LookupStringFromVarSub looks up a value from the predefined varsub.Source
// and writes the value to the pointer. An error is returned on type error.
// If the variable is not present, then nil is returned and the pointer is
// untouched.
```

File: pkg/wharfyml/visit/mapvisitor.go [hunk 15]
```diff
@@ -157,36 +168,45 @@ func (p MapVisitor) VisitStringWithVarSub(nodeKey, varLookup string, target *str
...
-	}
-	*target = b
-	return nil
+	return visitNode(p, key, target, Bool)
 }
 
+// LookupBoolFromVarSub looks up a value from the predefined varsub.Source
+// and writes the values to the pointer. An error is returned on type error.
+// If the variable is not present, then nil is returned and the pointer is
+// untouched.
```
Comment (Alexamakans) [inline pkg/wharfyml/visit/mapvisitor.go:196]:
```suggestion
// LookupBoolFromVarSub looks up a value from the predefined varsub.Source
// and writes the value to the pointer. An error is returned on type error.
// If the variable is not present, then nil is returned and the pointer is
// untouched.
```

File: pkg/wharfyml/visit/mapvisitor.go [hunk 16]
```diff
@@ -157,36 +168,45 @@ func (p MapVisitor) VisitStringWithVarSub(nodeKey, varLookup string, target *str
...
 func (p MapVisitor) LookupBoolFromVarSub(varLookup string, target *bool) error {
 	return lookupFromVarSub(p, varLookup, target, p.VisitBool)
 }
 
+// RequireBoolFromVarSub looks up a value from the predefined varsub.Source
+// and writes the values to the pointer. An error is returned if the looked up
+// variable is not found or on type errors.
```
Comment (Alexamakans) [inline pkg/wharfyml/visit/mapvisitor.go:203]:
```suggestion
// RequireBoolFromVarSub looks up a value from the predefined varsub.Source
// and writes the value to the pointer. An error is returned if the looked up
// variable is not found or on type errors.
```

File: pkg/wharfyml/visit/verify.go [hunk 17]
```diff
@@ -24,6 +28,9 @@ func VerifyTag(node *yaml.Node, wantStr string, wantTag string) error {
 	return nil
 }
 
+// VerifyKindAndTag checks if the given node has both the wanted kind and then
+// also then wanted tag. The wantStr is only used as a pretty string of the
+// wanted type in the potential error message.
```
Comment (Alexamakans) [inline pkg/wharfyml/visit/verify.go:33]:
```suggestion
// VerifyKindAndTag checks if the given node has the wanted kind and tag. The
// wantStr is only used as a pretty string of the wanted type in the potential
// error message.
```

File: pkg/wharfyml/visit/visit.go [hunk 18]
```diff
@@ -21,13 +21,15 @@ var (
 	ErrTooManyDocs      = errors.New("only 1 document is allowed")
 )
 
+// String will try read this YAML node as a string.
```
Comment (Alexamakans) [inline pkg/wharfyml/visit/visit.go:24]: `sed -i 's/try read/try to read/g' ./pkg/wharfyml/visit/visit.go`

**Review by Alexamakans — APPROVED** [review]
Very nice, good work 👍🏻

All of these are just minor grammar/typo related things I think, so I'll approve.
Sorry for all the multiples of `values` -> `value`,  but they were kind of case-by-case so I opted with suggesting on each of them.

File: pkg/wharfyml/visit/visit.go (hunk 18 above)
Comment (jilleJr) [reply pkg/wharfyml/visit/visit.go:24]: Cmon man, you're making it too easy :P

**Comment by jilleJr** [comment]
> Very nice, good work 👍🏻
> 
> All of these are just minor grammar/typo related things I think, so I'll approve. Sorry for all the multiples of `values` -> `value`, but they were kind of case-by-case so I opted with suggesting on each of them.

I liked that I could just click add to commit on all of them. As they all were self-contained suggestions I think it was better than fine

**Pull request merged**